### ISSUE TITLE
Handle multiple-line headers in twisted.web.http.HTTPClient

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2062]
|<img alt="kkinder's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| kkinder reported|
|-|-|
|Trac ID|trac#2062|
|Type|defect|
|Created|2006-09-05 23:16:49Z|
|Branch|https://github.com/twisted/twisted/tree/multi-line-headers-2062|

Many HTTP servers send multiple line headers, such as this:

```
Http-header: text
        (text-continued)
```

The header is continued with a newline and a tab.

Twisted's http.py does not handle this:

```python
        if line:
            key, val = line.split(':', 1)
            val = val.lstrip()
            self.handleHeader(key, val)
            if key.lower() == 'content-length':
                self.length = int(val)
```

Line, in this case, is only part of the header. I've attached a patch which addresses this problem.

Attachments:

* [http.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/329/329bccc75b40306e1ada36378bd690c9000f9c6c/de671f4e1a1c7ad95bd2e9d52024877656afaa1b.diff) (2069 bytes) - added by kkinder on 2006-09-05 23:17:27Z - Patch to have http.py handle multiple headers.
* [multiline-http-request-field.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/329/329bccc75b40306e1ada36378bd690c9000f9c6c/1a7cce7465e98322ae2d9aea4215eff2bb605ccf.patch) (4293 bytes) - added by jesstess on 2009-09-13 20:10:32Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2062 2062
type__defect defect
reporter__kkinder kkinder
priority__normal normal
milestone__ 
branch__branches_multi_line_headers_2062 branches/multi-line-headers-2062
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__web web
keywords__httpclient httpclient
time__1157498209000000 1157498209000000
changetime__1263773857000000 1263773857000000
version__None None
owner__ 
cc__jesstess cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kkinder's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|kkinder commented|
|-|-|

BTW, I recall having sent this in before, but I couldn't find a ticket for it, and it hasn't been fixed.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_4"></a>|@jesstess removed owner|
|-|-|

Thanks kkinder. As it turns out, RFCs 1945 (HTTP 1.0) and 2616 (HTTP 1.1) state that HTTP message header fields can span multiple lines if each extra line is preceded by at least one space or horizontal tab.

kkinder's patch broke web/test/test_http test cases and seemed to be using a lot of temp variables, so multiline-http-request-field.patch simplifies things and handles the space case as well. It also includes a test case in web/test/test_http, a doc string for HTTPClient.lineReceived as it is affected by the patch, and documentation for the instance variables for HTTPClient since I was adding one anyway (_header, which holds a possibly incomplete HTTP header).

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|z3p commented|
|-|-|

Some quick comments:

 1. Can you put this into a branch?
 1. `lineReceived()`s docstring says it parses the headers, but it also parses the initial status line.
 1. `flushHeader()`
  1. It should get broken out into an actual method.
  1. It should have a name that describes what it does; maybe `extractHeader()`?
  1. It shouldn't reference self._header directly; it should use the header argument passed in.  Alternatively, it shouldn't take an argument.
  1. It should have a test.
 1. `MultilineHeaders`
  1. The name should have `TestCase` in it.
  1. test_MultilineHeaders should be camelCased.
  1. expectedHeaders and ourHandleHeader need documentation.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess commented|
|-|-|

(In [27825]) Branching to 'multi-line-headers-2062'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess commented|
|-|-|

(In [27830]) Address review comments by z3p.

refs [#2062](https://github.com/twisted/twisted/issues/2062)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|TimAllen set owner to @jesstess|
|-|-|

When I check out this branch and run the tests, `twisted.web.test.test_proxy.ProxyClientTestCase.test_statusWithMessage` repeatably fails for me:
```
[ERROR]: twisted.web.test.test_proxy.ProxyClientTestCase.test_statusWithMessage

Traceback (most recent call last):
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/web/test/test_proxy.py", line 201, in test_statusWithMessage
    return self._testDataForward(
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/web/test/test_proxy.py", line 153, in _testDataForward
    client.dataReceived("\r\n" + body)
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/protocols/basic.py", line 251, in dataReceived
    why = self.lineReceived(line)
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/web/http.py", line 461, in lineReceived
    self.extractHeader(self._header)
  File "/mnt/aquaria/nobackup/home/tim/tmp/Twisted-git/twisted/web/http.py", line 424, in extractHeader
    key, val = header.split(':', 1)
exceptions.ValueError: need more than 1 value to unpack
```

Looks like `HTTPClient.lineReceived()` breaks if the response doesn't have any headers. Might be a useful thing to add to the HTTPClient unit-tests.

It kind of bugs me that HTTPClient returns the (arbitrary) whitespace used to mark continued header values as part of the header values. On the other hand, we aren't canonicalizing header names or decoding things to Unicode, so at least we're _consistently_ presenting raw-ish data.

Apart from that one test-failure, looks good to me.

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|TimAllen commented|
|-|-|

Oh, I nearly forgot: you'll need to add a description of the change to `twisted/web/topfiles/2062.bugfix` or perhaps `.news`.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_12"></a>|@jesstess commented|
|-|-|

(In [27866]) Handle header-less requests and strip leading whitespace from the
individual lines of multiline headers.

refs [#2062](https://github.com/twisted/twisted/issues/2062)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_13"></a>|@jesstess removed owner|
|-|-|

Review points by TimAllen addressed:
 * check that headers exist before trying to process them added in lineReceived
 * test_noHeaders added to cover that case
 * stripping the leading whitespace from the individual lines of the multiline headers
 * 2062.feature added in twisted/web/topfiles

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra set owner to @jesstess|
|-|-|

Can you get rid of oneliners like:

```
"""Convert chunk to string.

    @returns: tuple (result, remaining), may raise ValueError.
"""
```

as much as possible in surrounding code and apply PEP-8 standards as much as possible? People here tend to dislike fixing up a whole module like this for a single ticket but readability now should count as well..

Thanks for working on those old tickets, fresh Twisted == good.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_15"></a>|@jesstess commented|
|-|-|

(In [27882]) Bring docstring formatting in http.py up to the standard.

refs [#2062](https://github.com/twisted/twisted/issues/2062)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph set owner to @jesstess|
|-|-|

It's always good to fix bugs, but `twisted.web.client.Agent` is hopefully the wave of the future.  We should be deprecating stuff like `HTTPClient` as soon as possible, or, where it's a reasonable amount of effort to do so compatibly, re-implementing it in terms of the `_newclient` stuff.  This branch should probably land, but further changes to fix bugs in other `twisted.web.client` stuff should lean in the direction of adding deprecation warnings and pointing folks at the new APIs.

  1. The "Multiline headers are possible" comment in `lineReceived`'s docstring is unnecessary.  The docstring should not list every possible bug that might be covered: if we followed this example, it should also say "Headers which contain whitespace are accepted.  Headers which have a ":" in them are accepted.  Unknown status codes will be accepted.", etc etc.  It would be more helpful to draw attention to the fact that request _bodies_ will be parsed by something else (`rawDataReceived`).
  2. This is a bugfix, not a feature.  Multi-line headers are a required part of the HTTP spec.  Therefore, the file in topfiles should be .bugfix, not .feature.
  3. The test docstrings should be more descriptive.  Phrases like "can successfully do X" or "will not fail when X" should be avoided if possible.  For example, instead of "HTTPClient can successfully parse header-less requests", say, "An HTTP request with no headers will cause handleHeaders and handleEndHeaders to be called on HTTPClient subclasses".
  4. As I was writing that previous comment, I realized that the new tests don't make particularly good assertions.  They effectively only assert that `handleStatus` was called, not that handleEndHeaders was called, or even that any headers were received.  If `ourHandleHeader` is _never_ called, then its assertions will never be made.  Any test which asserts that a callback should be invoked really ought to assert that the callback was in fact called.
  5. The use of lineReceived rather than dataReceived is a little confusing; especially whe it comes to including the delimiter ('\r\n') in `lineReceived`'s argument, something that should never happen in real life.

Thanks for keeping up the maintenance work!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_18"></a>|@jesstess commented|
|-|-|

(In [27952]) The description of "unfolding" multiline headers in RFC 822 suggests
that the non-CRLF whitespace between lines should be preserved, so
change lineReceived to have that behavior and adjust the test cases
accordingly.

refs [#2062](https://github.com/twisted/twisted/issues/2062)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_19"></a>|@jesstess removed owner|
|-|-|

Points 1-5 from glyph's review are now addressed in the branch. The comment about CRLFs being passed to lineReceived sent me back to the RFCs looking for clarification on how whitespace between the lines of multiline headers is supposed to be handle, as TimAllen had brought this up previously in the ticket. RFC 2616 (HTTP/1.1) references RFC 822 (STANDARD FOR THE FORMAT OF ARPA INTERNET TEXT MESSAGES) regarding the format for multiline headers. Section 3.1.1. of that RFC says:

"The process of moving from this folded multiple-line representation of a header field to its single line representation is called "unfolding. Unfolding is accomplished by regarding CRLF immediately followed by a LWSP-char as equivalent to the LWSP-char."

which implies to me that leading whitespace on the continuing lines of multiline headers should be preserved in what's passed to extractHeader and then handleHeader. That's what's done now in the branch.

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_20"></a>|@moshez set owner to @jesstess|
|-|-|

In http.py:lineReceived:line 464:
Stylistically I would like it better if it was:
```
if not line:
    ...stuff under else...
    return
...stuff under if...
```
That would allow dedenting the stuff if/elif/else stuff currently there.

Thanks for changing the 1/0 stuff to True/False in the .firstLine stuff.

Overall, patch well done. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_21"></a>|@jesstess commented|
|-|-|

(In [28021]) Rework flow control in lineReceived.

refs [#2062](https://github.com/twisted/twisted/issues/2062)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_22"></a>|@jesstess set status to `closed`|
|-|-|

(In [28022]) Merge multi-line-headers-2062

Authors: kkinder, jesstess
Reviewers: z3p, TimAllen, thijs, glyph, moshez
Fixes: [#2062](https://github.com/twisted/twisted/issues/2062)

twisted.web.http.HTTPClient now supports multi-line headers, which are
a required part of the HTTP specification.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
